src/plugins/intel_cpu/src/nodes/non_max_suppression.cpp

```diff
@@ -685,193 +202,192 @@ void NonMaxSuppression::prepareParams() {
-    auto scoresMemPtr =  getChildEdgesAtPort(NMS_SELECTEDSCORES)[0]->getMemoryPtr();
-    const size_t validOutputs = std::min(filtBoxes.size(), maxNumberOfBoxes);
+    const size_t valid_outputs = std::min(start_offset, max_number_of_boxes);
 
-    if (!m_outStaticShape) {
-        VectorDims newDims{validOutputs, 3};
-        redefineOutputMemory({newDims, newDims, {1}});
-    }
+    if (m_defined_outputs[NMS_SELECTED_INDICES]) {
+        const size_t stride = 3lu;
+        const auto edges = getChildEdgesAtPort(NMS_SELECTED_INDICES);
+        auto memory = edges[0]->getMemoryPtr();
 
-    int selectedIndicesStride = indicesMemPtr->getDescWithType<BlockedMemoryDesc>()->getStrides()[0];
+        if (!m_out_static_shape) {
+            VectorDims new_shape{ valid_outputs, stride };
 
-    int *selectedIndicesPtr = reinterpret_cast<int *>(indicesMemPtr->getData());
-    float *selectedScoresPtr = reinterpret_cast<float *>(scoresMemPtr->getData());
+            const auto& curr_shape = memory->getDesc().getShape();
+            if (!curr_shape.isStatic() || curr_shape.getStaticDims() != new_shape) {
+                const auto new_mem_desc = getBaseMemDescAtOutputPort(NMS_SELECTED_INDICES)->cloneWithNewDims(new_shape, (valid_outputs == 0lu));
+                for (size_t i = 0lu; i < edges.size(); i++) {
+                    edges[i]->getMemoryPtr()->redefineDesc(new_mem_desc);
+                }
+            }
+        }
 
-    size_t idx = 0lu;
-    for (; idx < validOutputs; idx++) {
-        selectedIndicesPtr[0] = filtBoxes[idx].batch_index;
-        selectedIndicesPtr[1] = filtBoxes[idx].class_index;
-        selectedIndicesPtr[2] = filtBoxes[idx].box_index;
-        selectedIndicesPtr += selectedIndicesStride;
+        auto out_ptr = reinterpret_cast<int32_t *>(memory->getData());
+        int32_t* boxes_ptr = &(m_filtered_boxes[0].batch_index);
 
-        selectedScoresPtr[0] = static_cast<float>(filtBoxes[idx].batch_index);
-        selectedScoresPtr[1] = static_cast<float>(filtBoxes[idx].class_index);
-        selectedScoresPtr[2] = static_cast<float>(filtBoxes[idx].score);
-        selectedScoresPtr += selectedIndicesStride;
-    }
+        size_t idx = 0lu;
+        for (; idx < valid_outputs; idx++) {
+            memcpy(out_ptr, boxes_ptr, 12);
+            out_ptr += stride;
+            boxes_ptr += 4;
+        }
 
-    if (m_outStaticShape) {
-        std::fill(selectedIndicesPtr, selectedIndicesPtr + (maxNumberOfBoxes - idx) * selectedIndicesStride, -1);
-        std::fill(selectedScoresPtr, selectedScoresPtr + (maxNumberOfBoxes - idx) * selectedIndicesStride, -1.f);
+        if (m_out_static_shape) {
+            std::fill(out_ptr, out_ptr + (max_number_of_boxes - idx) * stride, -1);
+        }
     }
 
-    int *valid_outputs = reinterpret_cast<int *>(getChildEdgesAtPort(NMS_VALIDOUTPUTS)[0]->getMemoryPtr()->getData());
-    *valid_outputs = static_cast<int>(validOutputs);
-}
+    if (m_defined_outputs[NMS_SELECTED_SCORES]) {
```

Comment:
Can we add new method that will allow to redifine memory for corresponding otuput port: `redefineOutputMemory(size_t port, const VectorDims& newShape)`?
This new method can be reused here and inside original Node::redefineOutputMemory

Reply:
Done